### PR TITLE
Meta: migrate from travis-ci to Github Actions

### PR DESCRIPTION
**Pull request by ljharb**
(ref https://github.com/ljharb/object.assign/pull/81)

travis-ci's new pricing plan has effectively dropped support for open source repos, so we need to migrate off of it.

(To work properly, these workflows must be set to `pull_request_target`, not `pull_request`, and must have a GOOGLE_API_KEY secret available, as well as GH_TOKEN - still working on this)

### CODE REVIEW

**Comment by leobalter** [comment]
to make this work properly, you eventually need to manually disable the old expected checks from TravisCI or they will remain eternally waiting for status.

Disregard if you already know this, it's just trauma from an event in the past.

**Comment by ljharb** [comment]
Yes, as soon as this PR is ready to land, I'll make the actions checks "required" and the travis checks no longer required, but as soon as I do that, every PR will be unmergeable until they're rebased, so it's the last step before/during merge.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+name: 'ecma-262'
+
+on: [pull_request, push]
+
+jobs:
+  build:
+    name: 'build & lint ecmarkup'
+    runs-on: ubuntu-latest
```
Comment (septs) [inline .github/workflows/build.yml:8]:
```suggestion
    runs-on: ubuntu-20.04
```

> Ubuntu-20.04 is ready to be the default version for the “ubuntu-latest” label in GitHub Actions and Azure DevOps. This change will be rolled out over a period of several weeks beginning on November, 30. (December, 22 - January, 4 will be a migration break due to holidays)

from https://github.com/actions/virtual-environments/issues/1816

Comment (septs) [reply .github/workflows/build.yml:8]: image

Comment (ljharb) [reply .github/workflows/build.yml:8]: the ubuntu version should never make a difference for the majority of node project's CI, there's no benefit in changing from the default.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ljharb/actions/node/run@125357c
+        name: 'npm ci && npm run build-master'
+        with:
+          node-version: lts/*
+          use-npm-ci: true
+          command: 'build-master'
```
Comment (ExE-Boss) [inline .github/workflows/build.yml:17]: This should probably include the trailing newline.

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ljharb/actions/node/run@125357c
+        name: 'npm ci && npm run build-only ./scripts/auto-deploy.sh'
+        with:
+          node-version: lts/*
+          use-npm-ci: true
+          shell-command: |
+            npm run build-only && \
+            ./scripts/auto-deploy.sh
```
Comment (ExE-Boss) [inline .github/workflows/deploy.yml:23]: ditto

File: .github/workflows/preview.yml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      - uses: ljharb/actions/node/run@125357c
+        name: 'npm ci && npm run build-only && …'
+        env:
+          CI_PREVIEW_TOKEN: ${{ secrets.CI_PREVIEW_TOKEN }}
+          PULL_REQUEST: ${{ github.event.number }}
+        with:
+          node-version: lts/*
+          use-npm-ci: true
+          shell-command: |
+            npm run build-only && node scripts/insert_snapshot_warning && node scripts/publish-preview
```
Comment (ExE-Boss) [inline .github/workflows/preview.yml:22]: ditto

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ljharb/actions/node/install@56f291b2a0d9e3bce72634356ae949b1e054b9f8
+        name: 'nvm install lts/* && npm install'
+        with:
+          node-version: lts/*
+          use-npm-ci: true
+      - run: 'npm ci && npm run build-master'
```
Comment (septs) [inline .github/workflows/build.yml:17]: why not split?

```suggestion
      - run: npm ci
      - run: npm run build-master
```

Comment (ljharb) [reply .github/workflows/build.yml:17]: sure, thanks

**Comment by ljharb** [comment]
This is ready to go! See https://github.com/ljharb/ecma262/pull/13 for proof that the Preview status shows up on a PR (but that won't happen on this repo until after this PR is landed, and thus, won't ever happen on this PR).

Once this is landed / while landing it, i'll update the required branch protections so that any other PR must rebase in order to be mergeable.

File: .github/workflows/deploy.yml [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    if: ${{ github.repository == 'tc39/ecma262' }}
+
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ljharb/actions/node/install@56f291b2a0d9e3bce72634356ae949b1e054b9f8
+        name: 'nvm install lts/* && npm ci'
+        with:
+          node-version: lts/*
+          use-npm-ci: true
+      - run: npm run build-only ./scripts/auto-deploy.sh
```
Comment (bakkot) [inline .github/workflows/deploy.yml:21]: Does `npm run` allow you to specify multiple commands, or is this some other syntax I don't recognize?

Comment (ljharb) [reply .github/workflows/deploy.yml:21]: whoops, it is a typo. good catch! saves us a failed master build :-)

File: .github/workflows/preview.yml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        uses: actions/github-script@v3.1.0
+        with:
+          script: |
+            const { owner, repo } = context.repo;
+            const artifacts = await github.actions.listWorkflowRunArtifacts({
+               owner,
+               repo,
+               run_id: ${{ github.event.workflow_run.id }},
+            });
+            const [{ id: artifact_id }] = artifacts.data.artifacts.filter((artifact) => artifact.name == 'out');
```
Comment (bakkot) [inline .github/workflows/preview.yml:37]: Random nit:

```suggestion
            const { id: artifact_id } = artifacts.data.artifacts.find((artifact) => artifact.name == 'out');
```

**Review by bakkot — APPROVED** [review]
Hard to review this for correctness, since it interacts with a bunch of other systems, so take my approval as "this looks reasonable as far as I can tell, let's land it and see what happens".

File: .github/workflows/preview.yml (hunk 7 above)
Comment (ljharb) [reply .github/workflows/preview.yml:37]: ha, thanks. this was tweaked copy-pasta; i'll pull in your suggestion.

**Pull request merged**